File: apps/backend/internal/handlers/alerts.go
package handlers

import (
	"net/http"
	"time"

	"github.com/nilbyte/tallyoh/backend/internal/middleware"
	"github.com/nilbyte/tallyoh/backend/internal/money"
)

func (h *Handler) ListAlerts(w http.ResponseWriter, r *http.Request) {
	claims, ok := middleware.ClaimsFromContext(r.Context())
	if !ok {
		http.Error(w, `{"error":"unauthorized"}`, http.StatusUnauthorized)
		return
	}

	rows, err := h.db.Query(r.Context(), `
		SELECT id, budget_id, budget_name, threshold_pct, percent_used,
		       allocated_cents, spent_cents, is_read, created_at
		FROM budget_alerts
		WHERE user_id = $1
		ORDER BY created_at DESC
		LIMIT 50
	`, claims.UserID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	defer rows.Close()

	type alertItem struct {
		ID             string  `json:"id"`
		BudgetID       string  `json:"budgetId"`
		BudgetName     string  `json:"budgetName"`
		ThresholdPct   int     `json:"thresholdPct"`
		PercentUsed    float64 `json:"percentUsed"`
		Allocated      float64 `json:"allocated"`
		AllocatedCents int64   `json:"allocatedCents"`
		Spent          float64 `json:"spent"`
		SpentCents     int64   `json:"spentCents"`
		IsRead         bool    `json:"isRead"`
		CreatedAt      string  `json:"createdAt"`
	}

	items := make([]alertItem, 0)
	for rows.Next() {
		var a alertItem
		var createdAt time.Time
		if err := rows.Scan(
			&a.ID, &a.BudgetID, &a.BudgetName, &a.ThresholdPct, &a.PercentUsed,
			&a.AllocatedCents, &a.SpentCents, &a.IsRead, &createdAt,
		); err != nil {
			writeError(w, http.StatusInternalServerError, "internal error")
			return
		}
		a.Allocated = money.ToReais(a.AllocatedCents)
		a.Spent = money.ToReais(a.SpentCents)
		a.CreatedAt = createdAt.UTC().Format("2006-01-02T15:04:05Z")
		items = append(items, a)
	}

	writeJSON(w, http.StatusOK, map[string]any{"items": items})
}

func (h *Handler) MarkAlertRead(w http.ResponseWriter, r *http.Request) {
	claims, ok := middleware.ClaimsFromContext(r.Context())
	if !ok {
		http.Error(w, `{"error":"unauthorized"}`, http.StatusUnauthorized)
		return
	}

	id := r.PathValue("id")
	tag, err := h.db.Exec(r.Context(), `
		UPDATE budget_alerts SET is_read = true
		WHERE id = $1 AND user_id = $2
	`, id, claims.UserID)
	if err != nil || tag.RowsAffected() == 0 {
		writeError(w, http.StatusNotFound, "not found")
		return
	}

	w.WriteHeader(http.StatusNoContent)
}

func (h *Handler) ClearReadAlerts(w http.ResponseWriter, r *http.Request) {
	claims, ok := middleware.ClaimsFromContext(r.Context())
	if !ok {
		http.Error(w, `{"error":"unauthorized"}`, http.StatusUnauthorized)
		return
	}

	h.db.Exec(r.Context(), `
		DELETE FROM budget_alerts WHERE user_id = $1 AND is_read = true
	`, claims.UserID)

	w.WriteHeader(http.StatusNoContent)
}

File: apps/somma-expenses/backend/src/handlers/alerts.rs
use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::{IntoResponse, Response},
};
use chrono::{DateTime, Utc};
use serde::Serialize;
use serde_json::json;
use somma_common::{error_response, ok_json, to_reais, AuthClaims};
use sqlx::Row;
use uuid::Uuid;

use crate::AppState;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AlertItem {
    pub id: Uuid,
    pub budget_id: Uuid,
    pub budget_name: String,
    pub threshold_pct: i32,
    pub percent_used: f64,
    pub allocated: f64,
    pub allocated_cents: i64,
    pub spent: f64,
    pub spent_cents: i64,
    pub is_read: bool,
    pub created_at: String,
}

pub async fn list_alerts(
    State(state): State<AppState>,
    claims: AuthClaims,
) -> Response {
    let user_uuid = match Uuid::parse_str(&claims.user_id) {
        Ok(u) => u,
        Err(_) => return error_response(StatusCode::BAD_REQUEST, "invalid user id"),
    };

    let rows = match sqlx::query(
        r#"
        SELECT id, budget_id, budget_name, threshold_pct, percent_used,
               allocated_cents, spent_cents, is_read, created_at
        FROM budget_alerts
        WHERE user_id = $1
        ORDER BY created_at DESC
        LIMIT 50
        "#,
    )
    .bind(user_uuid)
    .fetch_all(&state.db)
    .await
    {
        Ok(r) => r,
        Err(_) => return error_response(StatusCode::INTERNAL_SERVER_ERROR, "internal error"),
    };

    let items: Vec<AlertItem> = rows
        .into_iter()
        .map(|r| {
            let created_at: DateTime<Utc> = r.get("created_at");
            let allocated_cents: i64 = r.get("allocated_cents");
            let spent_cents: i64 = r.get("spent_cents");
            AlertItem {
                id: r.get("id"),
                budget_id: r.get("budget_id"),
                budget_name: r.get("budget_name"),
                threshold_pct: r.get("threshold_pct"),
                percent_used: r.get("percent_used"),
                allocated: to_reais(allocated_cents),
                allocated_cents,
                spent: to_reais(spent_cents),
                spent_cents,
                is_read: r.get("is_read"),
                created_at: created_at.format("%Y-%m-%dT%H:%M:%SZ").to_string(),
            }
        })
        .collect();

    ok_json(json!({ "items": items }))
}

pub async fn mark_alert_read(
    State(state): State<AppState>,
    claims: AuthClaims,
    Path(id): Path<Uuid>,
) -> Response {
    let user_uuid = match Uuid::parse_str(&claims.user_id) {
        Ok(u) => u,
        Err(_) => return error_response(StatusCode::BAD_REQUEST, "invalid user id"),
    };

    let res = sqlx::query("UPDATE budget_alerts SET is_read = true WHERE id = $1 AND user_id = $2")
        .bind(id)
        .bind(user_uuid)
        .execute(&state.db)
        .await;

    match res {
        Ok(tag) if tag.rows_affected() > 0 => StatusCode::NO_CONTENT.into_response(),
        Ok(_) => error_response(StatusCode::NOT_FOUND, "not found"),
        Err(_) => error_response(StatusCode::INTERNAL_SERVER_ERROR, "internal error"),
    }
}

pub async fn clear_read_alerts(
    State(state): State<AppState>,
    claims: AuthClaims,
) -> Response {
    let user_uuid = match Uuid::parse_str(&claims.user_id) {
        Ok(u) => u,
        Err(_) => return error_response(StatusCode::BAD_REQUEST, "invalid user id"),
    };

    let _ = sqlx::query("DELETE FROM budget_alerts WHERE user_id = $1 AND is_read = true")
        .bind(user_uuid)
        .execute(&state.db)
        .await;

    StatusCode::NO_CONTENT.into_response()
}
